src/System.Private.CoreLib/shared/System/Text/DecoderNLS.cs

```diff
@@ -238,5 +240,188 @@ internal void ClearMustFlush()
+                    }
+                    else
+                    {
+                        goto DestinationTooSmall;
+                    }
+
+                case OperationStatus.NeedMoreData:
+                    if (MustFlush)
+                    {
+                        goto case OperationStatus.InvalidData; // treat as equivalent to bad data
+                    }
+                    else
+                    {
+                        persistNewCombinedBuffer = true;
+                        goto Finish; // successfully consumed some bytes, output no chars
+                    }
+
+                case OperationStatus.InvalidData:
+                    break;
+
+                default:
+                    Debug.Fail("Unexpected OperationStatus return value.");
+                    break;
+            }
+
+            // Couldn't decode the buffer. Fallback the buffer instead.
+
+            if (FallbackBuffer.Fallback(combinedBuffer.Slice(0, combinedBufferBytesConsumed).ToArray(), index: 0))
+            {
+                if (!_fallbackBuffer.TryDrainRemainingDataForGetChars(chars, out charsWritten))
+                {
+                    goto DestinationTooSmall;
+                }
+            }
+
+        Finish:
+
+            if (persistNewCombinedBuffer)
+            {
+                Debug.Assert(combinedBufferBytesConsumed == combinedBuffer.Length, "We should be asked to persist the entire combined buffer.");
+                SetLeftoverData(combinedBuffer); // the buffer still only contains partial data; a future call to Convert will need it
+            }
+            else
+            {
+                ClearLeftoverData(); // the buffer contains no partial data; we'll go down the normal paths
+            }
+
+            bytesConsumed = combinedBufferBytesConsumed - _leftoverByteCount; // amount of 'bytes' buffer consumed just now
+            return charsWritten;
+
+        DestinationTooSmall:
+
+            // If we got to this point, we're trying to write chars to the output buffer, but we're unable to do
+            // so. Unlike EncoderNLS, this type does not allow partial writes to the output buffer. Since we know
+            // draining leftover data is the first operation performed by any DecoderNLS API, there was no
+            // opportunity for any code before us to make forward progress, so we must fail immediately.
+
+            _encoding.ThrowCharsOverflow(this, nothingDecoded: true);
+            throw null; // will never reach this point
+        }
+
+        /// <summary>
+        /// Copies as many elements of <paramref name="srcLeft"/> followed by as many elements of <paramref name="srcRight"/>
+        /// as possible into <paramref name="dest"/>, then returns the total number of elements copied.
+        /// </summary>
+        private static int ConcatInto<T>(ReadOnlySpan<T> srcLeft, ReadOnlySpan<T> srcRight, Span<T> dest)
```

Comment:
Since this is operating on 4 bytes max, it would smaller and faster to just copy it byte-by-byte inline:

```
int total = 0;

for (int i = 0; i < srcLeft.Length; i++)
{
     if (total >= dest.Length)
         break;
     dest[total++] = srcLeft[i];
}

for (int i = 0; i < dstLeft.Length; i++)
{
     if (total >= dest.Length)
         break;
     dest[total++] = dstLeft[i];
}

return total;
```